### PR TITLE
Add analyzer: Avoid StringBuilder parameters for P/Invokes

### PR DESCRIPTION
**Pull request by elinor-fung**
Implementation of dotnet/runtime#35693

**ID**: CA1838
**Category**: Performance
**Severity**: Hidden, BulkConfigurable
**Default**: Disabled

cc @AaronRobinsonMSFT @jkoritzinsky

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/AvoidStringBuilderPInvokeParameters.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            var methodSymbol = (IMethodSymbol)context.Symbol;
+
+            // Only check P/Invokes
+            DllImportData dllImportData = methodSymbol.GetDllImportData();
+            if (dllImportData == null)
+            {
+                return;
+            }
+
+            // Warn on StringBuilder paramaters
```
Comment (AaronRobinsonMSFT) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/AvoidStringBuilderPInvokeParameters.cs:60]: ~~paramaters~~ -> parameters

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/AvoidStringBuilderPInvokeParameters.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidStringBuilderPInvokeParametersMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidStringBuilderPInvokeParametersDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        internal static DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+                                                        RuleId,
+                                                        s_localizableTitle,
+                                                        s_localizableMessage,
+                                                        DiagnosticCategory.Performance,
+                                                        RuleLevel.IdeHidden_BulkConfigurable, // Only for users explicitly targeting performance - addressing violation is non-trivial
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/AvoidStringBuilderPInvokeParameters.cs:27]: Thanks for the comment!

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx [hunk 3]
```diff
@@ -1422,4 +1422,13 @@
...
     <value>Do not use 'OutAttribute' on string parameters for P/Invokes</value>
   </data>
+  <data name="AvoidStringBuilderPInvokeParametersDescription" xml:space="preserve">
+    <value>Marshalling of StringBuilder always creates a native buffer copy, resulting in multiple allocations for one marshalling operation.</value>
+  </data>
+  <data name="AvoidStringBuilderPInvokeParametersMessage" xml:space="preserve">
+    <value>Avoid StringBuilder parameters for P/Invokes. Consider using a char buffer from ArrayPool instead</value>
+  </data>
+  <data name="AvoidStringBuilderPInvokeParametersTitle" xml:space="preserve">
+    <value>Avoid StringBuilder parameters for P/Invokes</value>
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1432]:
```suggestion
    <value>Avoid 'StringBuilder' parameters for P/Invokes</value>
```

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1432]: Same for other resource strings.

**Comment by elinor-fung** [comment]
I think this is ready to merge now.

**Comment by stephentoub** [comment]
Have you tried running this on dotnet/runtime? For the things it flags, how would you fix them?

**Pull request merged**